log/log.go

```diff
@@ -75,3 +72,13 @@ func Errorf(format string, v ...interface{}) {
 func Debugf(format string, v ...interface{}) {
 	debugf(format, v...)
 }
+
+// writer is a wrapper around harvester's logger that implements io.Writer.
+// It is available so we can keep using harvester's logger for external dependencies that require io.Writer (like hclog).
+type writer struct{}
+
+// Write log using log.Errorf() and will never returns an error.
+func (writer) Write(p []byte) (n int, err error) {
+	Errorf(string(p))
```

Comment:
This means that the writter will always log as an error which is not correct...
Previous consul implementation did not have an abstracted way to add an logger.
We should use existing logger facilities and pass it into consul in order to log in the format of the provided logger interface.

Reply:
As far as I can see hclog only uses this for Errors (see implementation of `pl.RunWithClientAndHclog` in `watcher.go`)

Reply:
In the PR you mentioned we used io.Writer directly (ie. os.Stderr or os.Stdout) which skips the logger that is used, more specifically if the service uses patron's logger the log output would be different and won't be parseable by fluentd  (as it won't be a JSON), this PR tries to address that by wrapping the logger with a io.Writer

Reply:
Seems odd because the Setup method passes in the writers (it should be patron) and the result should be loggerd using the patron logger which is structured. is this not the case?

Reply:
you can see the discussions here https://github.com/taxibeat/pricing-fare-service/pull/223#discussion_r466938326 and see how it is being used, we can in theory pass a writer and as the writer pass the logger (assuming the logger implements io.Writer) but why would we allow the user to define a separate writer and logger? why not define that internally every component will use the configured logger?

Reply:
The problem is that initially, consul did not support any logger at all, they just had their own which could not be changed.
If we can support creating a hclogger without using the writer, that solution would be ideal.
Passing in the writer (which is patron's writer should work as expected but is not ideal as you mentioned.
Looking at the docs of hclog there seems to be a way to do it by using: `hclog.SetDefault(log Logger)`.
We can implement the interface and us the pass in function of the setup method of harvester.
This approach is less hacky, it would support all log levels that might be introduced in the future and could also remove the need for the writer totally.
I think that this is a much better approach to support what we need. What do you think?

Reply:
I don't understand why consul even needs logging, usually libraries leave it for the users to decide on how to log, but since they do the only downside of letting them use the default (which is os.Stderr if I'm not mistaken) is that the logs will not be parseable (as they won't be json), that is the only downside - and this was the whole reasoning behind this PR.

I see why you feel this current approach is hacky, maybe because of the log levels?
If I compare the two options:
1. This PR:
 Advantage: All logs are written with the same structure and are parseable
 Disadvantage: consul will always use the error level (at the moment they only use errors but I guess you mean if in the future they introduce other types of logs)
2. Using the default logger
 Advantage: Consul can log with different levels - I couldn't think of other advantages, if you have any mind please share :) 
 Disadvantage: Using different loggers and different structures

Reply:
Not exactly. consul uses `hclog` which means it can use all levels eventually if it is not already.
The proposed approach is to implement the `hclog.Logger` interface and use harvesters log internally, which would be Patron in this case.
Advantages: All logs harvester plus consul will be logged in the same format. Future use of different levels will also be supported OOTB since we support everything. The log format will be whatever is provided in the setup of Harvester.
Disadvantages: None, except for doing the coding part since the interface is big...

Reply:
oh, ok, now I understand your idea, I misunderstood that you wanted to implement hclog.Logger.
I agree your approach is better, in theory this can be built upon this PR and it'll be easier to implement your approach (since it will be an internal change, just replacing the writer with a logger) but it does leave a tech debt regarding the error levels.

I'll leave it for your consideration if you want to proceed with this PR or reject it and wait for a more complete implementation with the Logger interface.